hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/IncrementalChunkBuffer.java

```diff
@@ -0,0 +1,266 @@
+
+  /** @return the i-th buffer if it exists; otherwise, return null. */
+  private ByteBuffer getAtIndex(int i) {
+    final ByteBuffer ith = i < buffers.size() ? buffers.get(i) : null;
+    if (ith != null) {
+      if (!isDuplicated) {
+        assertInt(getBufferCapacityAtIndex(i), ith.capacity(), "capacity", i);
+      } else {
+        final int limitIndex = limit / increment;
+        if (i < limitIndex) {
+          assertInt(increment, ith.capacity(), "capacity", i);
+        } else if (i == limitIndex) {
+          assertInt(getBufferCapacityAtIndex(i), ith.limit(), "capacity", i);
+        } else {
+          assertInt(0, ith.limit(), "capacity", i);
+        }
+      }
+    }
+    return ith;
+  }
+
+  /** @return the i-th buffer. It may allocate buffers. */
+  private ByteBuffer getAndAllocateAtIndex(int index) {
+    final int limitIndex = limit / increment;
+    // never allocate over limit
+    if (limit % increment == 0) {
+      Preconditions.checkArgument(index < limitIndex);
+    } else {
+      Preconditions.checkArgument(index <= limitIndex);
+    }
+
+    int size = buffers.size();
+    if (index < size) {
+      return getAtIndex(index);
+    }
+
+    // allocate upto the given index
+    ByteBuffer b = null;
+    for (; size <= index; size++) {
+      b = ByteBuffer.allocate(getBufferCapacityAtIndex(size));
+      buffers.add(b);
+    }
+    return b;
+  }
+
+  /** @return the buffer containing the position. It may allocate buffers. */
+  private ByteBuffer getAndAllocateAtPosition(int position) {
+    Preconditions.checkArgument(position < limit);
+    final int i = position / increment;
+    final ByteBuffer ith = getAndAllocateAtIndex(i);
+    assertInt(position%increment, ith.position(), "position", i);
+    return ith;
+  }
+
+  @Override
+  public int position() {
+   // The buffers list must be in the following orders:
+   // full buffers, buffer containing the position, empty buffers, null buffers
+    int i = 0;
+    ByteBuffer ith = null;
+    // buffers must be full
+    for (; i < buffers.size(); i++) {
+      ith = getAtIndex(i);
+      if (ith.position() != increment) {
+        break; // found the first non-full buffer
+      }
+    }
+    // buffer containing the position
+    final int position = i * increment + Optional.ofNullable(getAtIndex(i))
+        .map(ByteBuffer::position).orElse(0);
+    assertRemainingList(ith, i);
+    return position;
+  }
+
+  private void assertRemainingList(ByteBuffer ith, int i) {
+    if (ith != null) {
+      // buffers must be empty
+      for (i++; i < buffers.size(); i++) {
+        ith = getAtIndex(i);
+        if (ith == null) {
+          break; // found the first non-null
+        }
+        assertInt(0, ith.position(), "position", i);
+      }
+    }
+    // buffers must be null
+    for (i++; i < buffers.size(); i++) {
+      Preconditions.checkState(getAtIndex(i) == null);
+    }
+  }
+
+  @Override
+  public int remaining() {
+    return limit - position();
```

Comment:
maybe I am confused, but at this point, to compute 

> remaining

we will walk thru every buffer, since location the buffer with the current position needs iteration till the first empty buffer, but then we sanity check the rest of the list? Is my understanding correct ?

Reply:
remaining() here is the same as ByteBuffer.remaining().  Since we were refactored from ByteBuffer to ChunkBuffer, we keep having remaining().